blockchain/consensus/narwhal_cns/src/proposer.rs

```diff
@@ -0,0 +1,420 @@
+
+    let miner_addr = validator_addresses
+        .get(&validator_key)
+        .copied()
+        .ok_or(NarwhalConsensusError::NoMinerAddress(validator_key))?;
+
+    let consensus_index = output.consensus_output.consensus_index;
+
+    let mut messages = Vec::new();
+
+    for batch in output.transaction_batches {
+        for bytes in batch {
+            match SignedMessage::unmarshal_cbor(bytes.as_ref()) {
+                Ok(msg) => messages.push(msg),
+                Err(e) => {
+                    // Narwhal will let anything in, so a Byzantine validator can send us garbage.
+                    // There's no way for us to decide if the fault is with them, or our deseralizer.
+                    warn!("Error unmarshaling message: {}", e);
+                    continue;
+                }
+            };
+        }
+    }
+
+    create_block(state_manager, base, miner_addr, consensus_index, messages).await
+}
+
+/// Create a block from a Narwhal certificate and the included batches of transactions.
+async fn create_block<DB>(
+    state_manager: &Arc<StateManager<DB>>,
+    base: &Arc<Tipset>,
+    miner_addr: Address,
+    consensus_index: SequenceNumber,
+    messages: Vec<SignedMessage>,
+) -> anyhow::Result<GossipBlock>
+where
+    DB: BlockStore + Sync + Send + 'static,
+{
+    let smoke_height = state_manager.chain_config().epoch(Height::Smoke);
+    let (parent_state_root, parent_receipts) = state_manager.tipset_state(base).await?;
+    let parent_base_fee =
+        forest_chain::compute_base_fee(state_manager.blockstore(), base, smoke_height)?;
+
+    let parent_weight = NarwhalConsensus::weight(state_manager.blockstore(), base)?;
+    let persisted = forest_chain::persist_block_messages(
+        state_manager.blockstore(),
+        messages.iter().collect(),
+    )?;
+
+    // TODO: If we make a partial block then we have to keep track of the
+    // index of the last transaction we managed to include as well.
+    let index = ConsensusTransactionIndex {
+        consensus_index,
+        transactions_included: messages.len() as u32,
+        transactions_total: messages.len() as u32,
+    };
+
+    // Use the ticket to persist the consensus index, for crash recovery.
+    let ticket: Ticket = index.try_into()?;
+
+    // There is nothing in the certificate to suggest what time it was made, so we can't assign a timestamp.
+    let mut header = BlockHeader::builder()
+        .messages(persisted.msg_cid)
+        .bls_aggregate(Some(persisted.bls_agg))
+        .miner_address(miner_addr)
+        .ticket(Some(ticket))
+        .weight(parent_weight)
+        .parent_base_fee(parent_base_fee)
+        .parents(base.key().clone())
+        .epoch(base.epoch() + 1)
+        .state_root(parent_state_root)
+        .message_receipts(parent_receipts)
+        .build()?;
+
+    // We cannot sign the header because it has to be derived deterministically and identically on all nodes.
+    header.signature = None;
```

Comment:
Also an interesting challenge that Mir is facing. With all that we are learning from the integration of Mir and Narwhal we may be in a good place to abstract all of these pieces to integrate both.